components/raftstore-v2/src/worker/pd/slowness.rs

```diff
@@ -0,0 +1,154 @@
+                Duration::from_secs(0),
+                STORE_SLOW_TREND_RESULT_MISC_GAUGE_VEC.with_label_values(&["spike_filter_value"]),
+                STORE_SLOW_TREND_RESULT_MISC_GAUGE_VEC.with_label_values(&["spike_filter_count"]),
+                Duration::from_secs(120),
+                Duration::from_secs(15),
+                Duration::from_secs(60),
+                Duration::from_secs(300),
+                1,
+                2000,
+                STORE_SLOW_TREND_RESULT_MARGIN_ERROR_WINDOW_GAP_GAUGE_VEC
+                    .with_label_values(&["L1"]),
+                STORE_SLOW_TREND_RESULT_MARGIN_ERROR_WINDOW_GAP_GAUGE_VEC
+                    .with_label_values(&["L2"]),
+                cfg.slow_trend_unsensitive_result,
+            ),
+            slow_result_recorder: RequestPerSecRecorder::new(),
+            last_tick_finished: true,
+        }
+    }
+}
+
+impl<EK, ER, T> Runner<EK, ER, T>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+    T: PdClient + 'static,
+{
+    /// Record slowness periodically.
+    pub fn handle_update_slowness_stats(&mut self, _tick: u64, duration: RaftstoreDuration) {
+        self.slowness_stats.last_tick_finished = true;
+        // TODO: It's more appropriate to divide the factor into `Disk IO factor` and
+        // `Net IO factor`.
+        // Currently, to make the detection and judgement of Slowness of V2 compactible
+        // to V1, it summarizes all factors by `sum` simplily, approved valid to common
+        // cases when there exists IO jitters on Network or Disk.
+        self.slowness_stats.slow_cause.record(
+            tikv_util::time::duration_to_us(duration.sum()),
+            Instant::now(),
+        );
+    }
+
+    pub fn handle_slowness_stats_tick(&mut self) {
+        // The following code records a periodic "white noise", which helps
+        // mitigate any minor fluctuations in disk I/O or network I/O latency.
+        // After conducting extensive e2e testing, "100ms" has been determined
+        // to be the most suitable choice for it.
+        self.slowness_stats
+            .slow_cause
+            .record(100_000, Instant::now()); // 100ms
+        // Handle timeout if last tick is not finished as expected.
+        if !self.slowness_stats.last_tick_finished && self.is_store_heartbeat_delayed() {
+            // If the last slowness tick already reached abnormal state and was delayed for
+            // reporting by `store-heartbeat` to PD, we should report it here manually as a
+            // FAKE `store-heartbeat`. It's an assurance that the heartbeat to
+            // PD is not lost. Normally, this case rarely happens in
+            // raftstore-v2.
+            self.handle_fake_store_heartbeat();
+        }
+        // Move to next tick.
+        self.slowness_stats.last_tick_finished = false;
+    }
+
+    pub fn update_slowness_in_store_stats(&mut self, stats: &mut pdpb::StoreStats, query_num: u64) {
```

Comment:
It's better to add a unit test to cover this logic.

Reply:
IMO, it's redundant for it.
This logic is a compatible copy to `raftstore-v2` from `raftstore-v1`, and this test has already covered by the integration tests on `StoreHeartbeat`.

Reply:
@LykxSassinator If it's covered by v2's end-to-end test, that's fine. But it seems there's no integration test for slow score in either v1 or v2. 
I think you can think to add one in a separate PR.